### DISCUSSION TITLE
Is there a way to use Prometheus with Next.js?

### DISCUSSION THREAD

**Discussion by olamiwhat** [vercel/next.js#16205] [Help]
I am wondering if there is a way to integrate [Prometheus](https://prometheus.io/) with Next.js to scrape metrics. I know that [Next.js has a built-in relayer that allows you to analyze and measure the performance of pages using different metrics.](https://nextjs.org/docs/advanced-features/measuring-performance), just don't know how to hook that up with Prometheus. Any idea or help on this please?

**Comment by eric-hc** [comment]
Hey @olamiwhat, I am also interested in this. Were you able to find anything?

**Reply by weyert** [reply]
I haven't tried it out myself yet but I was planning to use `prom-client` (https://www.npmjs.com/package/prom-client) to send some metrics to Prometheus. My current understanding, still researching myself, is that Prometheus predominantly expects to pull metrics from endpoints instead that you push to Prometheus itself but I might have misunderstood that bit.

My current idea is to send metrics to Prometheus is to implement `reportWebVitals` which send it's metrics to a API route that I have defined e.g. `/api/quantify/page.ts` which main job is to collect metrics that will be send to Prometheus. Next, I would write a simple middleware for Next.js to catch the status codes, response times and send them to Prometheus too. 

I am not sure if I should collect the metrics somewhere and then just have a /metrics which include them all which Prometheus can poll or that I push the metrics myself directly or queue them and every n seconds push them.

**Reply by eric-hc** [reply]
`prom-client` would be able to expose server metrics, but that could work if you use your API route to collect the [client metrics](https://nextjs.org/docs/advanced-features/measuring-performance) for Prometheus to scrape. Have been thinking about this as well

**Reply by weyert** [reply]
Yeah, this all happily worked for me. Main blocker is that I can't get the route path (asPath?) in the custom server which causes a lot of useless metrics for each possible page e.g. `/posts/2323331` instead of `/posts/[id]`.

**Reply by jellebouwman** [reply]
@weyert stumbled upon this blog post that you might find interesting! https://danlaush.biz/posts/dynamic-routes-prometheus

**Reply by weyert** [reply]
@jellebouwman Cool, going to check it out :D

**Reply by hirehamir** [reply]
Do you have any links to blog posts or documentation that helped you integrate Prometheus with Next.js?

**Comment by maapteh** [comment]
In our cluster we are obliged to use it and normally we have it in express route INTERNAL handler as default. For next i have it fully working with the following setup below. The trick is using express as server, because we need custom other things as well :)

```
const { parse } = require('url');
const promClient = require('prom-client');
const next = require('next');
const express = require('express');
const dev = process.env.NODE_ENV !== 'production';

const HTTP_PORT = 3001;

const app = next({ dev });
const handle = app.getRequestHandler();
const server = express();

const registry = new promClient.Registry();

promClient.collectDefaultMetrics({
    register: registry,
});

const nextAppHandler = (req, res) => {
    const parsedUrl = parse(req.url, true);
    const { pathname } = parsedUrl;

    // assets folder in public which are own files
    if (pathname.startsWith('/assets')) {
        res.setHeader('Cache-Control', 'public, max-age=31536000, immutable');
    }

    handle(req, res, parsedUrl);

};


app.prepare().then(() => {
    // this is the endpoint we give for our metrics, in our cluster its never reachable from outside
    server.get('/metrics', (_, res) => {
        return res.send(registry.metrics());
    })
   
    server.all('*', (req, res) => {
        return nextAppHandler(req, res)
    })

    server.listen(HTTP_PORT, () => {
        console.log(
            `🚀 http application ready on http://localhost:${HTTP_PORT}`,
        );
    });

});
```

it then exposed the metrics:

```
# HELP process_cpu_user_seconds_total Total user CPU time spent in seconds.
# TYPE process_cpu_user_seconds_total counter
process_cpu_user_seconds_total 7.802911000000001

# HELP process_cpu_system_seconds_total Total system CPU time spent in seconds.
# TYPE process_cpu_system_seconds_total counter
process_cpu_system_seconds_total 0.8255330000000001

SNIP
```

**Reply by weyert** [reply]
Nice, I have been working on it too. The problem I am having that any pages of Next.js are counted as * in Prometheus. That's the last problem for me to solve.

Personally, not getting to excited about needing to use a custom server

**Reply by maapteh** [reply]
I see no other option, if you want to keep it in the "NextJS way" of working you need to make the endpoint public. And thats nice for simple sites (for example i see health check in public api, it works but thats it).

For routes and prometheus, i think (didnt see any router exposing in next) you just need the complete middleware for express https://www.npmjs.com/package/express-prometheus-middleware my team doesnt use it for route, just health of the node itself (using other metrics for the rest). Good luck.

**Reply by ivanlopes-jimdo** [reply]
Newbie question, but how did you get it to run? What does your dependencies look like? I can't get it to run, as prom-client complains about missing core Node modules, such as `cluster` and `child-process`.

**Reply by maapteh** [reply]
You run the express server to boot your next app? The above file is exactly what we use. Using latest next but worked with 8 for sure too. Maybe its possible to have it working in the Default api setup of next, but in my case i needed the express so didnt really look into it.

**Reply by ivanlopes-jimdo** [reply]
Just coming back here to update on the solution. I just realized Prometheus works flawlessly and out of the box on NextJS if you're using that inside your custom server, such as in this example, or inside the APIs. What me and my team were trying to do was actually monitor the time it took to run, for instance, getServerSideProps. When trying to do that we were struggling a lot, due to problems with importing it strictly in the server side.
The solution was to fallback to using Prometheus only on the custom server file, and monitor all incoming requests. `express-prom-bundle` is what we're using, as it makes the process really easy.

**Comment by kelly-tock** [comment]
is there any other way to do this without a custom server?  i'm also trying to avoid doing that.

**Reply by weyert** [reply]
Yeah, you could  probably use something like Opentelemetry to instrument http / express automatically by running it with `node -r opentelemetry-tracing.js` only the metrics part of Opentelemetry is still in the beta phase but appears to work alright for me. Only I haven't used `node -r` myself yet but I can iimagine it to work.

**Comment by gun0317** [comment]
@olamiwhat
this might help you - https://codesandbox.io/s/next-js-api-prom-client-hhvdi?file=/pages/api/metrics.ts

**Reply by gun0317** [reply]
In my case, I had to `await` the `register()`. Unless I got {}.
Minimal reproduction: https://codesandbox.io/s/next-js-api-prom-client-forked-ono9j?file=/pages/api/metrics.ts

**Reply by Moranilt** [reply]
> In my case, I had to `await` the `register()`. Unless I got {}. Minimal reproduction: https://codesandbox.io/s/next-js-api-prom-client-forked-ono9j?file=/pages/api/metrics.ts

this exactly the way to use it in NextJS. Thank you, it helps to me.

**Comment by weyert** [comment]
I am still struggling to get the route names back from Next e.g. `/pages/posts/[id]`. Anyone been able to find a way to do this?

**Comment by noveogroup-amorgunov** [comment]
Hey guys. How to add http custom metrics (requests per seconds, requests duration, 4xx, 5xx) in NextJS without custom server? I guess I need to use middleware API, but I don't know how to catch end of response. I don't see any callback or res.end in NextResponse

**Comment by hirehamir** [comment]
For anyone following this, Lee said

> We're focused on OpenTelemetry right now, but open to community contributions for examples [for Prometheus].
> 
> https://nextjs.org/docs/app/building-your-application/optimizing/open-telemetry

https://github.com/vercel/next.js/issues/56681#issuecomment-1802056337

**Comment by Netail** [comment]
With the new instrumentation hook & open-telementry, I was able to build my own prometheus exporter;

Enabling the experimental instrumentationHook
```js
// next.config.js
    experimental: {
        instrumentationHook: true,
        serverComponentsExternalPackages: [
            '@opentelemetry/instrumentation',
        ],
    }
```

Only use the instrumentation for nodejs environments (in my case)
```ts
// instrumentation.ts
export const register = async () => {
    if (process.env.NEXT_RUNTIME === 'nodejs') {
        await import('./instrumentation-node')
    }
};
```

Initialize OpenTelementry setup with some handy metrics and instrumentations
```ts
// instrumentation-node.ts
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { HostMetrics } from '@opentelemetry/host-metrics';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { RuntimeNodeInstrumentation } from '@opentelemetry/instrumentation-runtime-node';
import {
    Resource,
    detectResourcesSync,
    envDetector,
    hostDetector,
    processDetector,
} from '@opentelemetry/resources';
import { MeterProvider } from '@opentelemetry/sdk-metrics';
import {
    ATTR_SERVICE_NAME,
    ATTR_SERVICE_VERSION,
} from '@opentelemetry/semantic-conventions';

const exporter = new PrometheusExporter({
    port: 9464,
});
const detectedResources = detectResourcesSync({
    detectors: [envDetector, processDetector, hostDetector],
});

const customResources = new Resource({
    [ATTR_SERVICE_NAME]: 'my-app',
    [ATTR_SERVICE_VERSION]: '0.1.0',
});

const resources = detectedResources.merge(customResources);

const meterProvider = new MeterProvider({
    readers: [exporter],
    resource: resources,
});
const hostMetrics = new HostMetrics({
    name: `my-app-metrics`,
    meterProvider,
});

registerInstrumentations({
    meterProvider,
    instrumentations: [
        new HttpInstrumentation(),
        new RuntimeNodeInstrumentation(),
    ],
});

hostMetrics.start();
```

After start-up you can find the metrics at `localhost:9464/metrics`

<sup>Oh hello @maapteh, long time no see :D</sup>

**Reply by derekdowling** [reply]
I filed: https://github.com/open-telemetry/opentelemetry-js/issues/4890 to make this approach even more useful.

**Reply by ciandt-crodrigues** [reply]
@Netail Thanks for sharing man! 
It will help me a lot, but do you know if getting the metrics for each route/page is possible?
Do I need to create custom metrics for each of them even if I use this?

**Reply by Netail** [reply]
That's a good question, we would like to have this too, but haven't figured out yet. 
You have some plugins; (`@opentelemetry/instrumentation-http`, `@opentelemetry/instrumentation-router` & `@opentelemetry/instrumentation-express`) for e.g. node:http, router & express, but didn't seem to work with NextJS' routing as they use their own http server?
If you find a way, please let know :)

**Reply by Netail** [reply]
`@opentelemetry/instrumentation-http` seems to be getting the http.route from `getRPCMetadata(context.active())`, which needs to be set by nextjs if I understand https://github.com/open-telemetry/opentelemetry-js/issues/4890#issuecomment-2378959092 correctly...

**Reply by Netail** [reply]
Maybe @dvoytenko is able to help?

**Reply by ciandt-crodrigues** [reply]
@Netail
I was able to make it work with the following code:
``` typescript
import { context } from "@opentelemetry/api";
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks";
import { getRPCMetadata, RPCType, setRPCMetadata } from "@opentelemetry/core";
import { PrometheusExporter } from "@opentelemetry/exporter-prometheus";
import { HostMetrics } from "@opentelemetry/host-metrics";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { RuntimeNodeInstrumentation } from "@opentelemetry/instrumentation-runtime-node";
import {
  detectResourcesSync,
  envDetector,
  hostDetector,
  processDetector,
  Resource,
} from "@opentelemetry/resources";
import { MeterProvider } from "@opentelemetry/sdk-metrics";
import {
  ATTR_HTTP_ROUTE,
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} from "@opentelemetry/semantic-conventions";
import { IncomingMessage } from "http";

// manually setting a context manager to replace the no-op context manager
context.setGlobalContextManager(new AsyncLocalStorageContextManager());

const exporter = new PrometheusExporter({
  port: 9464,
});
const detectedResources = detectResourcesSync({
  detectors: [envDetector, processDetector, hostDetector],
});

const customResources = new Resource({
  [ATTR_SERVICE_NAME]: "MyApp",
  [ATTR_SERVICE_VERSION]: "0.1.0",
});

const resources = detectedResources.merge(customResources);

const meterProvider = new MeterProvider({
  readers: [exporter],
  resource: resources,
});
const hostMetrics = new HostMetrics({
  name: "olo-r:shell",
  meterProvider,
});

registerInstrumentations({
  meterProvider,
  instrumentations: [
    new HttpInstrumentation({
      requestHook: (span, request) => {
        const route = (request as IncomingMessage)?.url;
        if (route) {
          if (route && (route.endsWith(".json") || !route.includes("."))) {
            // Try to apply the route only for pages and client side fetches
            const rpcMetadata = getRPCMetadata(context.active()); // retrieve rpc metadata from the active context
            if (rpcMetadata) {
              if (rpcMetadata?.type === RPCType.HTTP) {
                rpcMetadata.route = route;
              }
            } else {
              setRPCMetadata(context.active(), {
                type: RPCType.HTTP,
                route,
                span,
              });
            }
          }
        }
      },
    }),
    new RuntimeNodeInstrumentation(),
  ],
});
hostMetrics.start();

```

this now includes the route

**Reply by rdsn** [reply]
Hi [Netail](https://github.com/Netail), I tried your solution and it works well in dev mode (next dev -p 8083), but in prod mode (next build && next start -p 8083) it lacks any info from instrumentation-http meaning that there are no http_server_duration_bucket metrics in http://localhost:9464/metrics (while they exist when running the app in dev mode). I am using next.js 14.2.17. Am I missing something? Which version do you use? Do you have http_server_duration_bucket metrics when running in prod mode?

**Reply by sadaffathali** [reply]
I have the same problem. any solution?

**Reply by Netail** [reply]
Awesome @ciandt-crodrigues, one remark; I would parse the query parameters off as it now also creates a counter for all `?_rsc=` requests

**Reply by QAston** [reply]
Same problem here, the HttpInstrumentation output doesn't show up in production, only in dev. Using next 14.0.4

**Reply by aizerin** [reply]
It works for me in production build with standalone output. I used the code above. I just adjusted `next.config.mjs` for nextjs 15.

```
const nextConfig = {
  output: 'standalone',
  serverExternalPackages: ['@opentelemetry/instrumentation'],
};

export default nextConfig;
```

**Reply by aizerin** [reply]
It is also working with @vercel/otel

```javascript
import { registerOTel } from '@vercel/otel';
import { envDetector, hostDetector, processDetector } from '@opentelemetry/resources';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { RuntimeNodeInstrumentation } from '@opentelemetry/instrumentation-runtime-node';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { getRPCMetadata, RPCType, setRPCMetadata } from '@opentelemetry/core';
import { context } from '@opentelemetry/api';

const exporter = new PrometheusExporter({
  port: 9464,
  endpoint: '/metrics',
  host: '0.0.0.0',
});

registerOTel({
  serviceName: 'app',
  resourceDetectors: [envDetector, processDetector, hostDetector],
  metricReader: exporter,
  instrumentations: [
    new HttpInstrumentation({
      requestHook: (span, request) => {
        const route = request?.url;
        if (route) {
          if (route && (route.endsWith('.json') || !route.includes('.'))) {
            const rpcMetadata = getRPCMetadata(context.active());
            if (rpcMetadata) {
              if (rpcMetadata?.type === RPCType.HTTP) {
                rpcMetadata.route = route;
              }
            } else {
              setRPCMetadata(context.active(), {
                type: RPCType.HTTP,
                route,
                span,
              });
            }
          }
        }
      },
    }),
    new RuntimeNodeInstrumentation(),
  ],
});
```

**Comment by Chimochikhin** [comment]
Any news about normal connection Prometheus to NextJs?

**Comment by flameddd** [comment]
I am completely new to Prometheus, but I think we can do something like this.

```js
import { NextApiResponse } from "next";
import client from "prom-client";

const collectDefaultMetrics = client.collectDefaultMetrics;
const Registry = client.Registry;
const register = new Registry();
collectDefaultMetrics({ register });


export default async function handle(_, res: NextApiResponse) {
  const metrics = await register.metrics()
  res.setHeader("Content-type", register.contentType);
  res.send(metrics);
};

```

## Is the result correct ?

<img width="1087" alt="image" src="https://github.com/vercel/next.js/assets/22259196/0e5d65cf-0ec7-4fd5-9260-2cfddc5e523b">

**Reply by victorhdchagas** [reply]
You tried to publish this on vercel? I have this logs locally, but on vercel still empty (Just two lines).

**Reply by flameddd** [reply]
> You tried to publish this on vercel? I have this logs locally, but on vercel still empty (Just two lines).

No, I never deploy this on vercel.

I did some search, look like `prom-client` did NOT support vercel egde. There are some discussion on https://github.com/siimon/prom-client/issues/584. I think some ppl found workaround, take a look if in need.

Good luck !!

**Reply by jhgeluk** [reply]
> I am completely new to Prometheus, but I think we can do something like this.
> 
> ```js
> import { NextApiResponse } from "next";
> import client from "prom-client";
> 
> const collectDefaultMetrics = client.collectDefaultMetrics;
> const Registry = client.Registry;
> const register = new Registry();
> collectDefaultMetrics({ register });
> 
> 
> export default async function handle(_, res: NextApiResponse) {
>   const metrics = await register.metrics()
>   res.setHeader("Content-type", register.contentType);
>   res.send(metrics);
> };
> ```
> 
> ## Is the result correct ?
> <img alt="image" width="1087" src="https://private-user-images.githubusercontent.com/22259196/342368660-0e5d65cf-0ec7-4fd5-9260-2cfddc5e523b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2Mzg5MzcsIm5iZiI6MTcyODYzODYzNywicGF0aCI6Ii8yMjI1OTE5Ni8zNDIzNjg2NjAtMGU1ZDY1Y2YtMGVjNy00ZmQ1LTkyNjAtMmNmZGRjNWU1MjNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDA5MjM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1Y2QwYmYxNDc5OGU2NmM1NTJhMDQ3N2UyZGIxMjEzZWI1NDZiNmM1YjlmYWJhNTY1ZmYyY2Y3Nzg2OTQ0MGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Yumslvw-HWzVPwPifpjvZF_ytZ-y4qBN3tLrr_vlFRk">

Where did you place this code @flameddd

**Reply by flameddd** [reply]
> Where did you place this code @flameddd

https://nextjs.org/docs/pages/building-your-application/routing/api-routes

------

For `App Router` you need to slightly modify my example code

**Reply by flameddd** [reply]
After few months, I realized that my solution is significantly different with [Express.js solution](https://github.com/vercel/next.js/discussions/16205#discussioncomment-142021).

My solution only provides basic information, like CPU or memory.  

On the other hand, `Expressjs` collects information for all routes. If you need to gather data on all APIs or endpoints, you should use the `Expressjs` solution.

I haven't tried the OpenTelemetry solution yet, but it seems similar to the `Expressjs` one.

cc @victorhdchagas  @jhgeluk 

thanks for all

**Comment by jhgeluk** [comment]
Is there any way we can collect and create metrics out of the default spans provided by next.js? I'm missing the "route/path" information in the "http_server_duration_bucket" metrics.

**Comment by heilgar** [comment]
here is solution I found working on next 15+ (at least what I have now): 
https://gist.github.com/heilgar/08ae697b8b279dbd0a80c58eef5a6cad
